src/pybind/mgr/cephadm/services/smb.py

```diff
@@ -31,15 +31,34 @@ def ranked(self, spec: ServiceSpec) -> bool:
         smb_spec = cast(SMBSpec, spec)
         return 'clustered' in smb_spec.features
 
+    def fence(self, daemon_id: str) -> None:
+        logger.info(f'Fencing old smb.{daemon_id}')
+        ret, out, err = self.mgr.mon_command({
+            'prefix': 'auth rm',
+            'entity': f'client.smb.{daemon_id}',
+        })
+
     def fence_old_ranks(
         self,
         spec: ServiceSpec,
         rank_map: Dict[int, Dict[int, Optional[str]]],
         num_ranks: int,
     ) -> None:
-        logger.warning(
-            'fence_old_ranks: Unsupported %r %r', rank_map, num_ranks
-        )
+        for rank, m in list(rank_map.items()):
+            if rank >= num_ranks:
+                for daemon_id in m.values():
+                    if daemon_id is not None:
+                        self.fence(daemon_id)
+                del rank_map[rank]
+                self.mgr.spec_store.save_rank_map(spec.service_name(), rank_map)
+            else:
+                max_gen = max(m.keys())
+                for gen, daemon_id in list(m.items()):
+                    if gen < max_gen:
+                        if daemon_id is not None:
+                            self.fence(daemon_id)
+                        del rank_map[rank][gen]
+                        self.mgr.spec_store.save_rank_map(spec.service_name(), rank_map)
```

Comment:
Just to capture I comment I had in a meeting:
Because this code is so similar to what nfs is doing, and the fact that the system adds things to the rank map outside of the `CephService` subclasses I think we should, in the future, refactor this part of the rank map management.

We can probably replace the `fence_old_ranks` function with a common method in the base class (or elsewhere) and then provide hooks for the `run_grace_tool` method and the like for the service specific pieces. -- But this is all for a later cleanup effort IMO.